packages/js-sdk/src/user/loginWithMFA.ts

```diff
@@ -0,0 +1,168 @@
+import isObjectLike from 'lodash/isObjectLike';
+import isBoolean from 'lodash/isBoolean';
+import { KinveyError } from '../errors/kinvey';
+import {
+  setSession,
+  setMFASessionToken,
+  removeMFASessionToken,
+  formatKinveyBaasUrl,
+  HttpRequestMethod,
+  KinveyHttpRequest,
+  KinveyBaasNamespace,
+  KinveyHttpAuth,
+  setDeviceToken,
+  getDeviceToken,
+  hasDeviceToken,
+  removeDeviceToken,
+} from '../http';
+import { User } from './user';
+import { LoginOptions, validateNoActiveUser, executeLoginRequest, validateCredentials } from './login';
+
+const errMsgNoAuthenticators = 'User has no MFA authenticators yet. Please enable MFA.';
+
+export interface MFAContext {
+  mfaSessionToken: string;
+  retries: number;
+  authenticator?: any;
+  error?: any;
+}
+
+async function executeChallengeRequest(authenticatorId: string): Promise<any> {
+  const request = new KinveyHttpRequest({
+    method: HttpRequestMethod.POST,
+    auth: KinveyHttpAuth.MFASessionToken,
+    url: formatKinveyBaasUrl(KinveyBaasNamespace.User, '/mfa/challenge'),
+    body: { authenticatorId },
+  });
+
+  const response = await request.execute();
+  return response.data;
+}
+
+async function executeCompleteRequest(code, trustDevice = false): Promise<object> {
+  const request = new KinveyHttpRequest({
+    method: HttpRequestMethod.POST,
+    auth: KinveyHttpAuth.MFASessionToken,
+    url: formatKinveyBaasUrl(KinveyBaasNamespace.User, '/mfa/complete'),
+    body: {
+      code,
+      createDeviceToken: trustDevice,
+    },
+  });
+
+  const response = await request.execute();
+  response.data.user._kmd.authtoken = response.data.authToken;
+  return response.data;
+}
+
+async function completeMFALoginRetryable(
+  mfaComplete: (authenticator: string, context: MFAContext) => Promise<MFACompleteResult>,
+  context: MFAContext
+): Promise<any> {
+  const errMsgNoCode = 'MFA code is missing.';
+  const errMsgTrustDevice = 'trustDevice should be boolean.';
+  try {
+    const mfaCompleteResult = await mfaComplete(context.authenticator.id, context);
+    if (!isObjectLike(mfaCompleteResult) || mfaCompleteResult.code == null) {
+      throw new KinveyError(errMsgNoCode);
+    }
+
+    if (mfaCompleteResult.trustDevice != null && !isBoolean(mfaCompleteResult.trustDevice)) {
+      throw new KinveyError(errMsgTrustDevice);
+    }
+    const trustDevice = mfaCompleteResult.trustDevice || false;
+    const mfaData = await executeCompleteRequest(mfaCompleteResult.code, trustDevice);
+    return mfaData;
+  } catch (err) {
+    if (err.message === errMsgNoCode || err.message === errMsgTrustDevice || !(err instanceof KinveyError)) {
```

Comment:
Moving mfaComplete outside of the try block will remove this check and have the same behaviour.